src/scanner.c

```diff
@@ -0,0 +1,442 @@
+#include "tag.h"
+#include <wctype.h>
+
+enum TokenType {
+    START_TAG_NAME,
+    SCRIPT_START_TAG_NAME,
+    STYLE_START_TAG_NAME,
+    END_TAG_NAME,
+    ERRONEOUS_END_TAG_NAME,
+    SELF_CLOSING_TAG_DELIMITER,
+    IMPLICIT_END_TAG,
+    RAW_TEXT,
+    COMMENT
+};
+
+typedef struct {
+    uint32_t len;
+    uint32_t cap;
+    Tag *data;
+} tags_vec;
+
+typedef struct {
+    tags_vec tags;
+} Scanner;
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+#define VEC_RESIZE(vec, _cap)                                                  \
```

Comment:
might be useful to factorize those in a separate vector.h? And reuse the same file across those different tree-sitter-xxx?
Otherwise each time we'll find a bug in one of those macros we will need to update 20 repositories ...

Reply:
I wouldn't mind refactoring this logic into a simple api for all tree-sitter parsers to use such that they don't have to rely on C++ - that being said, and not to toot my own horn but these have been extensively tested & fuzzed across many languages and many times for segfaults/overflows/leaks - I do think they are sound

Reply:
That's even more reason to factor them out and make sure other parsers (not only inside the tree-sitter org) use them!